src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -326,108 +312,73 @@ private BoundExpression BindSwitchExpression(SwitchExpressionSyntax node, Bindin
+        {
+            var bindingDiagnostics = BindingDiagnosticBag.GetInstance(diagnostics);
 
-                        case BoundIndexOrRangePatternIndexerAccess boundIndexOrRangePatternIndexerAccess:
-                            lengthProperty = boundIndexOrRangePatternIndexerAccess.LengthOrCountProperty;
-                            patternSymbol = boundIndexOrRangePatternIndexerAccess.PatternSymbol;
-                            found = true;
-                            break;
+            if (inputType.IsDynamic())
+            {
+                Error(bindingDiagnostics, ErrorCode.ERR_UnsupportedTypeForListPattern, node, inputType);
+            }
 
-                        case var v:
-                            throw ExceptionUtilities.UnexpectedValue(v.Kind);
-                    }
-                    analyzedArguments.Free();
-                    indexerGroup.Free();
-                    goto done;
+            receiverPlaceholder = new BoundListPatternReceiverPlaceholder(node, GetValEscape(inputType, inputValEscape), inputType) { WasCompilerGenerated = true };
+            bool hasErrors = false;
+            if (inputType.IsSZArray())
+            {
+                hasErrors |= !TryGetSpecialTypeMember(Compilation, SpecialMember.System_Array__Length, node, bindingDiagnostics, out PropertySymbol lengthProperty);
+                if (lengthProperty is not null)
+                {
+                    lengthAccess = new BoundPropertyAccess(node, receiverPlaceholder, lengthProperty, LookupResultKind.Viable, lengthProperty.Type) { WasCompilerGenerated = true };
+                }
+                else
+                {
+                    lengthAccess = new BoundBadExpression(node, LookupResultKind.Empty, ImmutableArray<Symbol?>.Empty, ImmutableArray<BoundExpression>.Empty, CreateErrorType(), hasErrors: true) { WasCompilerGenerated = true };
                 }
-                lookupResult.Clear();
             }
+            else
+            {
+                hasErrors |= !TryBindLengthOrCount(node, receiverPlaceholder, out lengthAccess, bindingDiagnostics);
+            }
+
+            var analyzedArguments = AnalyzedArguments.GetInstance();
+            var systemIndexType = GetWellKnownType(WellKnownType.System_Index, bindingDiagnostics, node);
+            argumentPlaceholder = new BoundListPatternIndexPlaceholder(node, systemIndexType) { WasCompilerGenerated = true };
+            analyzedArguments.Arguments.Add(argumentPlaceholder);
+
+            indexerAccess = BindElementAccessCore(node, receiverPlaceholder, analyzedArguments, bindingDiagnostics).MakeCompilerGenerated();
+            indexerAccess = CheckValue(indexerAccess, BindValueKind.RValue, bindingDiagnostics);
+            Debug.Assert(indexerAccess is BoundIndexerAccess or BoundIndexOrRangePatternIndexerAccess or BoundArrayAccess or BoundBadExpression or BoundDynamicIndexerAccess);
```

Comment:
>Debug.Assert(indexerAccess is BoundIndexerAccess or BoundIndexOrRangePatternIndexerAccess or BoundArrayAccess or BoundBadExpression or BoundDynamicIndexerAccess); [](http://example.com/codeflow?start=12&length=162)

It looks like the assert doesn't match what comment says for BoundSlicePattern.IndexAcceess